crates/bevy_reflect/src/type_data.rs

```diff
@@ -0,0 +1,63 @@
+use downcast_rs::{impl_downcast, Downcast};
+
+/// A trait for types generated by the [`#[reflect_trait]`][0] attribute macro.
+///
+/// [0]: crate::reflect_trait
+pub trait TypeData: Downcast + Send + Sync {
+    fn clone_type_data(&self) -> Box<dyn TypeData>;
+}
+impl_downcast!(TypeData);
+
+impl<T: 'static + Send + Sync> TypeData for T
+where
+    T: Clone,
+{
+    fn clone_type_data(&self) -> Box<dyn TypeData> {
+        Box::new(self.clone())
+    }
+}
+
+/// Trait used to generate [`TypeData`] for trait reflection.
+///
+/// This is used by the `#[derive(Reflect)]` macro to generate an implementation
+/// of [`TypeData`] to pass to [`TypeRegistration::insert`].
+pub trait FromType<T> {
```

Comment:
I think it would also be worthwhile to reference the other struct somewhere in the documentation here, briefly indicating how they can be used together.

Reply:
are you thinking something like "to specialize an existing blanket implementation, see `SpecializedFromType`"?

Reply:
Exactly 🙂